src-self-hosted/stage1.zig

```diff
@@ -527,3 +528,388 @@ export fn stage2_progress_update_node(node: *std.Progress.Node, done_count: usiz
+    if (cpu_str == null) return null;
+    if (arch_str == null) return null;
+
+    const arch = Target.parseArchTag(std.mem.toSliceConst(u8, arch_str.?)) catch {
+        return null;
+    };
+    return parseCpu(arch, std.mem.toSliceConst(u8, cpu_str.?)) catch |err| {
+        switch (err) {
+            error.OutOfMemory => @panic("out of memory"),
+            else => return null,
+        }
+    };
+}
+
+// ABI warning
+export fn stage2_target_details_parse_features(arch_str: ?[*:0]const u8, features_str: ?[*:0]const u8) ?*Stage2TargetDetails {
+    if (features_str == null) return null;
+    if (arch_str == null) return null;
+    
+    const arch = Target.parseArchTag(std.mem.toSliceConst(u8, arch_str.?)) catch return null;
+    return parseFeatures(arch, std.mem.toSliceConst(u8, features_str.?)) catch |err| {
+        switch (err) {
+            error.OutOfMemory => @panic("out of memory"),
+            else => return null,
+        }
+    };
+}
+
+fn parseCpu(arch: @TagType(std.Target.Arch), str: []const u8) !*Stage2TargetDetails {
+    const allocator = std.heap.c_allocator;
+
+    const cpus = std.target.getCpusForArch(arch);
+
+    for (cpus) |cpu| {
+        if (std.mem.eql(u8, str, cpu.name)) {
+            const ptr = try allocator.create(Stage2TargetDetails);
+            ptr.* = try Stage2TargetDetails.initCpu(std.heap.c_allocator, arch, cpu);
+
+            return ptr;
+        }
+    }
+
+    return error.InvalidCpu;
+}
+
+fn parseFeatures(arch: @TagType(std.Target.Arch), str: []const u8) !*Stage2TargetDetails {
+    const allocator = std.heap.c_allocator;
+
+    const known_features = std.target.getFeaturesForArch(arch);
+
+    var features = std.ArrayList(*const std.target.Feature).init(allocator);
+    defer features.deinit();
+
+    var start: usize = 0;
+    while (start < str.len) {
+        const next_comma_pos = std.mem.indexOfScalar(u8, str[start..], ',') orelse str.len - start;
+        const feature_str = std.mem.trim(u8, str[start..start+next_comma_pos], " ");
+
+        start += next_comma_pos + 1;
+
+        if (feature_str.len == 0) continue;
+
+        var feature: ?*const std.target.Feature = null;
+        for (known_features) |known_feature| {
+            if (std.mem.eql(u8, feature_str, known_feature.name)) {
+                feature = known_feature;
+                break;
+            }
+        }
+
+        if (feature) |f| {
+            features.append(f) catch @panic("out of memory");
+
+        } else {
+            return error.InvalidFeature;
+        }
+    }
+
+    const features_slice = features.toOwnedSlice();
+
+    const ptr = try allocator.create(Stage2TargetDetails);
+    ptr.* = try Stage2TargetDetails.initFeatures(allocator, arch, features_slice);
+
+    return ptr;
+}
+
+// ABI warning
+export fn stage2_target_details_get_cache_str(target_details: ?*const Stage2TargetDetails) [*:0]const u8 {
+    if (target_details) |td| {
+        return @as([*:0]const u8, switch (td.target_details) {
```

Comment:
This `@as` is redundant, it's implied by the fact that the return type is `[*:0]const u8`